File: sdks/kotlin/hearth-core/src/test/kotlin/io/hearth/sdk/PkceTest.kt
package io.hearth.sdk

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotEquals
import kotlin.test.assertTrue

class PkceTest {

    @Test
    fun `generatePkce returns verifier of at least 43 characters`() {
        val pair = generatePkce()
        // 32 bytes → base64url without padding = 43 chars
        assertTrue(pair.verifier.length >= 43, "verifier length was ${pair.verifier.length}")
    }

    @Test
    fun `generatePkce challenge is SHA-256 of verifier base64url-encoded without padding`() {
        val pair = generatePkce()
        val digest = java.security.MessageDigest.getInstance("SHA-256")
        val hash = digest.digest(pair.verifier.toByteArray())
        val expected = java.util.Base64.getUrlEncoder().withoutPadding().encodeToString(hash)
        assertEquals(expected, pair.challenge)
    }

    @Test
    fun `generatePkce method is always S256`() {
        val pair = generatePkce()
        assertEquals("S256", pair.method)
    }

    @Test
    fun `generatePkce returns different pairs on each call`() {
        val pair1 = generatePkce()
        val pair2 = generatePkce()
        assertNotEquals(pair1.verifier, pair2.verifier)
        assertNotEquals(pair1.challenge, pair2.challenge)
    }

    @Test
    fun `generatePkce verifier contains only base64url characters without padding`() {
        val pair = generatePkce()
        assertTrue(
            pair.verifier.all { it.isLetterOrDigit() || it == '-' || it == '_' },
            "verifier contained invalid chars: ${pair.verifier}"
        )
        assertTrue(!pair.verifier.contains('='), "verifier must not be padded")
    }

    @Test
    fun `generatePkce challenge contains only base64url characters without padding`() {
        val pair = generatePkce()
        assertTrue(
            pair.challenge.all { it.isLetterOrDigit() || it == '-' || it == '_' },
            "challenge contained invalid chars: ${pair.challenge}"
        )
        assertTrue(!pair.challenge.contains('='), "challenge must not be padded")
    }
}

File: sdks/node/src/pkce.test.ts
/** §PKCE — generatePkce() tests (TDD — written before implementation). */

import { describe, it, expect } from "vitest";
import { createHash } from "node:crypto";
import { generatePkce } from "./pkce.js";

describe("generatePkce", () => {
  it("returns a PkcePair with verifier, challenge, and method S256", () => {
    const pair = generatePkce();
    expect(typeof pair.verifier).toBe("string");
    expect(typeof pair.challenge).toBe("string");
    expect(pair.method).toBe("S256");
  });

  it("verifier is 43 Base64url characters (32 bytes, no padding)", () => {
    const { verifier } = generatePkce();
    expect(verifier).toHaveLength(43);
  });

  it("verifier contains only Base64url-safe characters", () => {
    const { verifier } = generatePkce();
    expect(verifier).toMatch(/^[A-Za-z0-9\-_]+$/);
  });

  it("verifier has no padding characters", () => {
    const { verifier } = generatePkce();
    expect(verifier).not.toContain("=");
  });

  it("challenge is BASE64URL(SHA256(verifier)) with no padding", () => {
    const { verifier, challenge } = generatePkce();
    const expected = createHash("sha256").update(verifier).digest("base64url");
    expect(challenge).toBe(expected);
    expect(challenge).not.toContain("=");
  });

  it("challenge is 43 Base64url characters (SHA-256 = 32 bytes → 43 chars)", () => {
    const { challenge } = generatePkce();
    expect(challenge).toHaveLength(43);
  });

  it("successive calls produce unique pairs (CSPRNG)", () => {
    const p1 = generatePkce();
    const p2 = generatePkce();
    expect(p1.verifier).not.toBe(p2.verifier);
    expect(p1.challenge).not.toBe(p2.challenge);
  });

  it("method is always 'S256'", () => {
    for (let i = 0; i < 5; i++) {
      expect(generatePkce().method).toBe("S256");
    }
  });
});
